### PR TITLE
Feature/issue 262 cvodes

### PR DESCRIPTION
**Pull request by wds15**
#### Submisison Checklist

- [X] Run unit tests: `./runTests.py test/unit`
- [X] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:
Replace CVODE ODE integrator with CVODES.

#### Intended Effect:
Make ODE integrator more robust and faster.

#### How to Verify:
Tests for CVODE have been rewritten for CVODES. So look at the code and run the ODE related tests.

#### Documentation:
doxygen comments are written, may need more.

#### Reviewer Suggestions: 
@bob-carpenter @betanalpha @syclik 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is requiring  to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bob-carpenter** [comment]
prim/arra/functor/cvodes_integrator.hpp

* What comes out of that err_handler and why silence it?

* Strings should always be passed by reference --- there's a `func_name` that's being apssed by value.  And you shouldn't have an "_" at the end of a function name.

* Why is there a solver_ == 1 and solver_ == 2 test before creating memory?  What other options are there?  Do you wnat to throw an exception if the solver is something else?

* Do the comparison with NULL if that's what you're checking for cvode_mem_

* You're setting memory to the solver class, which seems odd (but maybe the comment is wrong)

* Can you consolidate the solver branching logic either up front or at the end?  It's easy to get wrong when it's spread out and also hard to follow.  I don't like the way this is combining solvers.  Why not have two different ones rather than all the branching with everything shared pulled into a subclass?

* is the S_ > 0 comparision really an != NULL comparison?

* Why is there a double* argument to rhs?  Is this required by some other interface?  Otherwise, we shouldn't be using pointers anywhere --- everything should be done by reference.  It makes the coding way easier (none of the this-> business) and it makes it more reliable (no segfaulting when they don't exist).

* I take it the return 0 in ode_rhs is required?  It seems pointless given there's no other return value.

* Same comment about raw pointers in rhs_sens --- are these really necessary?  

* Why is there a reference to AD in rhs_sens?  This is in prim.   So shouldn't all of this go in a different file because it depends on autodiff?

* You don't need to doc what Eigen::Map does --- assume the reader of your code knows the external classes and what they do.  And don't add that extra padded space.

* I take it there's not a way to map an entire matrix and simplify that ySdot_eig calc?

* `for (size_t n = 0; n < N_; n++) state[n] = y0_dbl_[n];` can be done using `assign` with iterators.

I think I'm going to hold off on the rest until we can discuss this bit.

**Comment by wds15** [comment]
Thanks a lot for taking a first look, your comments cleared a lot of things up for me. More inline:

> Am 08.04.2016 um 23:29 schrieb Bob Carpenter <notifications@github.com>:
> 
> prim/arra/functor/cvodes_integrator.hpp
> 
> 	• What comes out of that err_handler and why silence it?

The errors are silenced, since we repack them into proper C++ exceptions which is the Stan way to report problems (same as CVODE code)

> 
> 	• Strings should always be passed by reference --- there's a func_name that's being apssed by value. And you shouldn't have an "_" at the end of a function name.

Ok, changed.

> 
> 	• Why is there a solver_ == 1 and solver_ == 2 test before creating memory? What other options are there? Do you wnat to throw an exception if the solver is something else?

Ok, I am throwing now an exception if solver_ is set to anything else other than 0, 1 or 2.

> 
> 	• Do the comparison with NULL if that's what you're checking for cvode_mem_

Yes, you are right. I changed to NULL as you point out. However, we may have to revisit this with C++11 which defines NULL differently as C.

> 
> 	• You're setting memory to the solver class, which seems odd (but maybe the comment is wrong)

What specific do you mean. However, I am almost sure that what you comment on just looks bad from a C++ perspective, but this is the way CVODES must be initialized.

> 
> 	• Can you consolidate the solver branching logic either up front or at the end? It's easy to get wrong when it's spread out and also hard to follow. I don't like the way this is combining solvers. Why not have two different ones rather than all the branching with everything shared pulled into a subclass?

Yeah, I see your point and tried to consolidate as good as possible. The thing is that things have to be done in correct order for CVODE. Following your proposal to have the branching logic for solver_ in one place I introduced a set_cvode_options function, because the order must be: first init memory, then set options and finally initialize the dense solver for the stiff case only. Also note, that CVODES is on purpose programmed as an extension to CVODE which is why the first steps are CVODE related and only the part which is additional due to the forward sensitivity calculations are CVODES related (which is why I named the function set_cvode_options).

> 
> 	• is the S_ > 0 comparision really an != NULL comparison?

I do prefer to think in terms of S_, but you are right, one can do a != NULL comparison here as well. Changed.

> 
> 	• Why is there a double* argument to rhs? Is this required by some other interface? Otherwise, we shouldn't be using pointers anywhere --- everything should be done by reference. It makes the coding way easier (none of the this-> business) and it makes it more reliable (no segfaulting when they don't exist).

I transformed this to a c array. One could transform things in the ode_rhs function into C++ vectors and pass these into the rhs function, if that’s preferable. However, this can only be done for the vectors y and dy_dt, we will still have some arrays of NV_Vector around (for the sensitivities).

> 
> 	• I take it the return 0 in ode_rhs is required? It seems pointless given there's no other return value.

Good point. In fact, CVODES allows the RHS function to return
different values. That is 0 for success, a positive value for a
recoverable error. A negative value for a non-recoverable error. A
positive recoverable error can be the case when the user knows that
the solution must be positive, but the numerics of CVODES tries to
evaluate the RHS at a negative state y due to (potentially)
linearization issues. In short, an option allowing the RHS function
in Stan to indicate such a failure would be good to have. I guess this
could be done a special exception which can be thrown in the user
rhs? It would be a useful feature I think.

> 
> 	• Same comment about raw pointers in rhs_sens --- are these really necessary?

I am afraid, these are necessary, particularly the N_Vector *yS things. This is the way CVODES handles it’s internal data structures and I don’t see another way to tightly integrate with these these. So N_Vector *yS is an array of N_Vector (there are S_ times NV_Vectors of size N_).

> 
> 	• Why is there a reference to AD in rhs_sens? This is in prim. So shouldn't all of this go in a different file because it depends on autodiff?

This comments makes the logic much clearer to me, I did not know this. You are right, things need to be moved which I did for most of my code to live now in „rev“.

> 
> 	• You don't need to doc what Eigen::Map does --- assume the reader of your code knows the external classes and what they do. And don't add that extra padded space.

Ok and I do think without the extra padding it is worse to read, but I removed.

> 
> 	• I take it there's not a way to map an entire matrix and simplify that ySdot_eig calc?

No. CVODES gives me an array of N_Vectors, but not a continuous chunk of memory.

> 
> 	• for (size_t n = 0; n < N_; n++) state[n] = y0_dbl_[n]; can be done using assign with iterators.

I think assign would resize the the lhs to the same size as the rhs which I do not want. I replace the for loop by a copy call which does not resize the vector.

> 
> I think I'm going to hold off on the rest until we can discuss this bit.

Great comments. I hope to have addressed them accordingly. Two questions from my side:

- Tests do currently handle the stiff solver only - I would propose to
  test stiff and non-stiff where it makes sense, i.e. precision tests,
  others?

- Should the large hornberg model as stiff performance test ( `test/unit/math/rev/arr/functor/integrate_ode_perf_test.cpp`) be
  included or only the smaller the smaller coupled_mm ( `test/unit/math/rev/arr/functor/integrate_ode_perf_small_test.cpp`)?

Sebastian

**Comment by betanalpha** [comment]
Just a comment — all of those pointers are to be compatible with the C callbacks.

**Comment by bob-carpenter** [comment]
> On Apr 9, 2016, at 6:16 AM, wds15 <notifications@github.com> wrote:
> 
> Thanks a lot for taking a first look, your comments cleared a lot of things up for me. More inline:
> 
> > Am 08.04.2016 um 23:29 schrieb Bob Carpenter <notifications@github.com>:
> > 
> > prim/arra/functor/cvodes_integrator.hpp
> > 
> > • What comes out of that err_handler and why silence it?
> 
> The errors are silenced, since we repack them into proper C++ exceptions which is the Stan way to report problems (same as CVODE code)

But how do we get the messages to repack them?


> > • Strings should always be passed by reference --- there's a func_name that's being apssed by value. And you shouldn't have an "_" at the end of a function name.
> 
> Ok, changed.
> 
> > 
> > • Why is there a solver_ == 1 and solver_ == 2 test before creating memory? What other options are there? Do you wnat to throw an exception if the solver is something else?
> 
> Ok, I am throwing now an exception if solver_ is set to anything else other than 0, 1 or 2.
> 
> > 
> > • Do the comparison with NULL if that's what you're checking for cvode_mem_
> 
> Yes, you are right. I changed to NULL as you point out. However, we may have to revisit this with C++11 which defines NULL differently as C.

I hadn't realized.  If you need to make it 0 to keep it compatible going forward,
it should be done that way everywhere.

> > 
> > • You're setting memory to the solver class, which seems odd (but maybe the comment is wrong)
> 
> What specific do you mean. However, I am almost sure that what you comment on just looks bad from a C++ perspective, but this is the way CVODES must be initialized.

I see that everyting is just calling cvode_mem_.  Does that mean
that cvode_mem_ isn't just memory, but it's the memory for all
of CVODE?  In which case, I'd just call it cvode_ in the code.
That's where my confusion was.  

> 
> 
> > • Can you consolidate the solver branching logic either up front or at the end? It's easy to get wrong when it's spread out and also hard to follow. I don't like the way this is combining solvers. Why not have two different ones rather than all the branching with everything shared pulled into a subclass?
> 
> Yeah, I see your point and tried to consolidate as good as possible. The thing is that things have to be done in correct order for CVODE. Following your proposal to have the branching logic for solver_ in one place I introduced a set_cvode_options function, because the order must be: first init memory, then set options and finally initialize the dense solver for the stiff case only.

Great.  That's exactly what I was hoping would be possible.

> Also note, that CVODES is on purpose programmed as an extension to CVODE which is why the first steps are CVODE related and only the part which is additional due to the forward sensitivity calculations are CVODES related (which is why I named the function set_cvode_options).
> 
> > 
> > • is the S_ > 0 comparision really an != NULL comparison?
> 
> I do prefer to think in terms of S_, but you are right, one can do a != NULL comparison here as well. Changed.

S_ is declared as a size_t, so it can't be negative.  
But if it's something that indicates a size, then S_ > 0 is
right;  if it's being used as a poitner somewhere, then != 0
or != NULL is the right thing to do.  I'm afraid you had this
right the first time now that I look at it again.

> 
> > • Why is there a double* argument to rhs? Is this required by some other interface? Otherwise, we shouldn't be using pointers anywhere --- everything should be done by reference. It makes the coding way easier (none of the this-> business) and it makes it more reliable (no segfaulting when they don't exist).
> 
> I transformed this to a c array. One could transform things in the ode_rhs function into C++ vectors and pass these into the rhs function, if that’s preferable. However, this can only be done for the vectors y and dy_dt, we will still have some arrays of NV_Vector around (for the sensitivities).

Thanks for the clarification.  

> > 
> > • I take it the return 0 in ode_rhs is required? It seems pointless given there's no other return value.
> 
> Good point. In fact, CVODES allows the RHS function to return
> different values. That is 0 for success, a positive value for a
> recoverable error. A negative value for a non-recoverable error. A
> positive recoverable error can be the case when the user knows that
> the solution must be positive, but the numerics of CVODES tries to
> evaluate the RHS at a negative state y due to (potentially)
> linearization issues. In short, an option allowing the RHS function
> in Stan to indicate such a failure would be good to have. I guess this
> could be done a special exception which can be thrown in the user
> rhs? It would be a useful feature I think.

Perhaps --- I'm not sure what it does.  I just thought a function that
had a return that was always the same looked fishy.

> 
> > 
> > • Same comment about raw pointers in rhs_sens --- are these really necessary?
> 
> I am afraid, these are necessary, particularly the N_Vector *yS things. This is the way CVODES handles it’s internal data structures and I don’t see another way to tightly integrate with these these. So N_Vector *yS is an array of N_Vector (there are S_ times NV_Vectors of size N_).
> 

OK.
> 
> > • Why is there a reference to AD in rhs_sens? This is in prim. So shouldn't all of this go in a different file because it depends on autodiff?
> 
> This comments makes the logic much clearer to me, I did not know this. You are right, things need to be moved which I did for most of my code to live now in „rev“.

Thanks.  Keeping this stuff organized is a huge pain, but
it helps downstream.

> > 
> > • You don't need to doc what Eigen::Map does --- assume the reader of your code knows the external classes and what they do. And don't add that extra padded space.
> 
> Ok and I do think without the extra padding it is worse to read, but I removed.

Thanks.  I find it easier to read if everything's always spaced
the same way.

> > 
> > • I take it there's not a way to map an entire matrix and simplify that ySdot_eig calc?
> 
> No. CVODES gives me an array of N_Vectors, but not a continuous chunk of memory.

Too bad.

> > 
> > • for (size_t n = 0; n < N_; n++) state[n] = y0_dbl_[n]; can be done using assign with iterators.
> 
> I think assign would resize the the lhs to the same size as the rhs which I do not want. I replace the for loop by a copy call which does not resize the vector.

I see.  This is why it's hard to refactor other people's code!

> > 
> > I think I'm going to hold off on the rest until we can discuss this bit.
> 
> Great comments. I hope to have addressed them accordingly. Two questions from my side:
> 
> - Tests do currently handle the stiff solver only - I would propose to
> test stiff and non-stiff where it makes sense, i.e. precision tests,
> others?

Yes, please do.

> 
> - Should the large hornberg model as stiff performance test ( `test/unit/math/rev/arr/functor/integrate_ode_perf_test.cpp`) be
> included or only the smaller the smaller coupled_mm ( `test/unit/math/rev/arr/functor/integrate_ode_perf_small_test.cpp`)?

If you could keep the tests under a minute, that'd be great.  If they're
testing the same thing, then we only need one.

Thanks!

- Bob

**Comment by wds15** [comment]
> > The errors are silenced, since we repack them into proper C++ exceptions which is the Stan way to report problems (same as CVODE code)
> 
> But how do we get the messages to repack them?

The integrate_times function sticks all calls to CVode into check_flag. The check_flag function fires an exception if and only if the return code from CVode is smaller than 0 (which is a non-recoverable error). This way the user gets to know the error code from CVode (which is packed into the exception message) and he get’s to know which CVode function trigger this. This information is enough to look up in the CVode manual what happened.

> > Yes, you are right. I changed to NULL as you point out. However, we may have to revisit this with C++11 which defines NULL differently as C.
> 
> I hadn't realized. If you need to make it 0 to keep it compatible going forward,
> it should be done that way everywhere.

I double checked things and using NULL is the correct thing to do; C+11 has defined NULL as nullptr, but this object is implicitly convertible to 0 in this case. In short C++11 is designed to work in this case as expected.

> 
> > > 
> > > • You're setting memory to the solver class, which seems odd (but maybe the comment is wrong)
> > 
> > What specific do you mean. However, I am almost sure that what you comment on just looks bad from a C++ perspective, but this is the way CVODES must be initialized.
> 
> I see that everyting is just calling cvode_mem_. Does that mean
> that cvode_mem_ isn't just memory, but it's the memory for all
> of CVODE? In which case, I'd just call it cvode_ in the code.
> That's where my confusion was. 

If you don’t mind I would leave it named cvode_mem_ named as is. This is the name as used in the cvodes user manual. The cvode_mem_ area is where cvodes stores its internal data structures used by the solver.

> 
> > I do prefer to think in terms of S_, but you are right, one can do a != NULL comparison here as well. Changed.
> 
> S_ is declared as a size_t, so it can't be negative. 
> But if it's something that indicates a size, then S_ > 0 is
> right; if it's being used as a poitner somewhere, then != 0
> or != NULL is the right thing to do. I'm afraid you had this
> right the first time now that I look at it again.
> 

No, you were right in changing this. I.e. suppose the user uses a wrong solver argument, then S_ will be set non-zero, but the memory won’t be initialized. That’s a case which should never happen in the framework of a Stan program, but using NULL as being right now is the right thing, I think.
> 

> > Good point. In fact, CVODES allows the RHS function to return
> > different values. That is 0 for success, a positive value for a
> > recoverable error. A negative value for a non-recoverable error. A
> > positive recoverable error can be the case when the user knows that
> > the solution must be positive, but the numerics of CVODES tries to
> > evaluate the RHS at a negative state y due to (potentially)
> > linearization issues. In short, an option allowing the RHS function
> > in Stan to indicate such a failure would be good to have. I guess this
> > could be done a special exception which can be thrown in the user
> > rhs? It would be a useful feature I think.
> 
> Perhaps --- I'm not sure what it does. I just thought a function that
> had a return that was always the same looked fishy.

I think adding the possibility to return >0 error codes from the RHS is a great thing, but we can add this feature later.
> 

> > - Tests do currently handle the stiff solver only - I would propose to
> > test stiff and non-stiff where it makes sense, i.e. precision tests,
> > others?
> 
> Yes, please do.

Ok, I extended the precision tests for stiff and non-stiff.

> 
> > 
> > - Should the large hornberg model as stiff performance test ( `test/unit/math/rev/arr/functor/integrate_ode_perf_test.cpp`) be
> > included or only the smaller the smaller coupled_mm ( `test/unit/math/rev/arr/functor/integrate_ode_perf_small_test.cpp`)?
> 
> If you could keep the tests under a minute, that'd be great. If they're
> testing the same thing, then we only need one.

Ok, I removed the horn berg example. BTW, it is really amazing to see the difference between a stiff and a non-stiff solver on this tiny michaelis-menten example! A non-stiff solver takes ages and the stiff-solver gives the result in ~200ms. Quite cool.

Finally I checked the code for cpplint, ran all header-tests and I declared all those rhs, rhs_sens and dense_jacobian function to be const functions as they do not change the object at all.

Any remarks on ode_model.hpp or decouple_states.hpp?

Sebastian

**Comment by bob-carpenter** [comment]
Thanks for the clarification on errors, NULL, and naming.  Sounds good.

I'll take a look at the other files you mention ASAP.

**Comment by syclik** [comment]
Hi Sebastian, I'm finally able to look at it. I reread through the existing comments. Here are mine.

1. `stan/math/rev/arr/functor/coupled_ode_system_cvode.hpp` was changed to `stan/math/rev/arr/functor/cvodes_integrator.hpp`. What was the reason for the name change?
2. This change looks like it propagated to `stan/math/rev/arr/functor/integrate_ode_cvode.hpp`. Can you summarize the change? I'm having getting my head around the change.
3. We can remove the cvode code if we're going forward with this. I believe we decided not to keep both CVODE and CVODES.

I'm a bit tired now, so I'll try to look at it again tomorrow. I'm running this command to tell what files have changed:
```
> git diff --name-only develop -- stan
stan/math/rev/arr.hpp
stan/math/rev/arr/functor/coupled_ode_system_cvode.hpp
stan/math/rev/arr/functor/cvodes_integrator.hpp
stan/math/rev/arr/functor/decouple_states.hpp
stan/math/rev/arr/functor/integrate_ode_cvode.hpp
stan/math/rev/arr/functor/integrate_ode_cvodes.hpp
stan/math/rev/arr/functor/ode_model.hpp
```

**Comment by wds15** [comment]
Hi Daniel,

great that you look into this!

On Tue, Apr 12, 2016 at 5:20 AM, Daniel Lee <notifications@github.com>
wrote:

> Hi Sebastian, I'm finally able to look at it. I reread through the
> existing comments. Here are mine.
>
>    1. stan/math/rev/arr/functor/coupled_ode_system_cvode.hpp was changed
>    to stan/math/rev/arr/functor/cvodes_integrator.hpp. What was the
>    reason for the name change?
>
> The logic of how things work is changed now. So before Stan was wrapping
around the base ODE system and expanded it to become the coupled ODE
system. Then the coupled ODE system was given to the integrator. However,
CVODES knows itself of how to calculate the sensitivities, all what we have
to do from the Stan side is to enable this integrator functionality and
feed into CVODES the required input. This is why I changed to name to
cvodes_integrator as there is not anymore a coupled ODE system on the Stan
side, instead we feed into a sensitivity enabled integrator the ODE RHS and
the RHS of the sensitivity system.

Since coupled_ode_system_cvode.hpp included already all the code needed to
interface with the C code, it made perfect sense to use it as a starting
base.


>
>    1. This change looks like it propagated to
>    stan/math/rev/arr/functor/integrate_ode_cvode.hpp. Can you summarize
>    the change? I'm having getting my head around the change.
>
> Not sure what you refer to. In any case, we should settle on how the
functionality is exposed to the language. I.e. the current proposal is to
remove integrate_ode_cvode and replace it with 4 flavors of
integrate_ode_cvodes (stiff/non-stiff simple/extended interface). However,
it was left open so far to what extend the language definitions can be used
to generate those function signatures. Maybe we quickly discuss later today.


>
>    1. We can remove the cvode code if we're going forward with this. I
>    believe we decided not to keep both CVODE and CVODES.
>
> Yes, CVODE is to be removed and my branch did replace CVODE with CVODES
right away. There is really no point in keeping the old one.

**Comment by bob-carpenter** [comment]
@wds15  This is now failing because of dependencies.  As we discussed, you just need to move anything that depends on autodiff to the rev directory to get beyond these dependency errors from Jenkins.

**Comment by wds15** [comment]
I checked and the problems seems not to be with rev, but with arr. Guessing from the error message

stan/math/rev/arr/functor/cvodes_integrator.hpp(237): File includes an Eigen header. [arr]

gives me the impression that I should move stuff as

stan/math/rev/arr  => stan/math/rev/mat

where Eigen headers are allowed apparently (my guess). I hope this is correct... should I move the respective tests as well?

Sebastian

**Comment by syclik** [comment]
Yes. That's correct. But why does it depend on Eigen? I didn't trace through the code closely enough to catch the dependency.

**Comment by wds15** [comment]
Using Eigen allows me to use linear algebra expressions and (most importantly) gives me the convenient Eigen::Map facility. The Map facility in turn is very handy to map memory chunks from CVODES to Eigen structures. Ultimately I can write involved expression much more comprehensible (matrix vector products) and the integration with CVODES is much tighter and faster (much reduced need to copy stuff around). However, the interface to Stan is still using a vector based representation. So Eigen data structures are used internally while to the outside things are done as before using vectors.

If that's fine, then I move things from arr to mat.

**Comment by bob-carpenter** [comment]
Yes, fine to move things to mat.

- Bob

**Comment by wds15** [comment]
Done... dear Jenkins please be happy!

**Comment by syclik** [comment]
We don't need to worry about this failure now. It's caused by Stan not finding the cvodes header. I'm going to review the code once more.

**Comment by syclik** [comment]
- I've removed `integrate_ode_cvode.hpp`.
- in `integrate_ode_cvodes.hpp`, why is there a typedef for `cvodes_integrator` on line 109? I think the indirection makes things unclear.
- `decouple_states`:
    1. if it's to stay a standalone function, it's in the wrong directory. It's not a functor, but just a function, so should be in the `fun` directory.
    2. If it's a standalone function, I would prefer it be called something more specific, like `decouple_ode_states`.
    3. My preference is for this to be a member function of `cvodes_integrator` because I don't believe it's meant to be used outside of `cvodes_integrator`.
    4. An alternative would be for `integrate_times` to decouple the states and return the proper values. I think that may be the best option.
-  `cvodes_integrator.hpp`
    - class needs doc
    - is `report_stats()` necessary? If so, can you point me to the cvodes doc for it? if not, can we remove the code?
    - more doc on the functions
    - is `sens` a common abbreviation for sensitivity? If not, could we just write it out? Code that's easier to read is helpful.
    - in the constructor, could we check the `solver` int values before moving forward? Perhaps using `check_bounded` would be more appropriate. I think that first if statement could be rewritten so it's clearer.
    - rather than using `boost::is_same`, we have template metaprograms for `is_const` or `is_const_struct` which indicates whether the type is constant or not.
    - why are we passing in double values for `y0` and `theta` instead of their natural types? This would simplify the type logic in `cvodes_integrator`.

Ok... before I go on, I think I need to step back and see what you're trying to do. I don't understand why there's so much indirection. `integrate_ode_cvodes` calls `cvodes_integrator` which then calls `ode_model`... to pass in values to `cvodes_integrator`, values are converted to double and things are controlled with template parameters instead of passing in the natural types, etc.

I think it's in the right direction, but I'm not sure it's where it needs to be to replace cvode. I can spent time tomorrow trying to figure it out, but perhaps we should split this into multiple pull requests:
1. add cvodes library
2. add `integrate_ode_cvodes` (this will be a much smaller diff and we can make line-level comments if we split it this way)
3. remove `integrate_ode_cvode` and cvode library

@wds15, thoughts?

**Comment by wds15** [comment]
Hi Daniel,

On Thu, Apr 14, 2016 at 7:25 AM, Daniel Lee <notifications@github.com>
wrote:

>
>    - I've removed integrate_ode_cvode.hpp.
>
> Yes, this function can be removed

>
>    - in integrate_ode_cvodes.hpp, why is there a typedef for
>    cvodes_integrator on line 109? I think the indirection makes things
>    unclear.
>
> By now this is unnecessary, I agree. I will remove it.

>
>    - decouple_states:
>       1. if it's to stay a standalone function, it's in the wrong
>       directory. It's not a functor, but just a function, so should be in the
>       fun directory.
>       2. If it's a standalone function, I would prefer it be called
>       something more specific, like decouple_ode_states.
>       3. My preference is for this to be a member function of
>       cvodes_integrator because I don't believe it's meant to be used
>       outside of cvodes_integrator.
>       4. An alternative would be for integrate_times to decouple the
>       states and return the proper values. I think that may be the best option.
>
> I would suggest then to rename to decouple_ode_states and move to the
right place. This function is intended to be used by other integrators
which also do forward mode sensitivity calculation. That is, the RK45
integrator could be refactored to use this function as well to avoid code
duplication.

>
>    -
>       - cvodes_integrator.hpp
>       - class needs doc
>
> Sure, I can add more; as far as I understand each member function also
needs some doc.

>
>    -
>       - is report_stats() necessary? If so, can you point me to the
>       cvodes doc for it? if not, can we remove the code?
>
> report_stats is not necessary at all to carry out the integration. It is
useful when you profile things. I found it useful, but if you want to keep
things to a bare minimum, then it should be removed.

>
>    -
>       - more doc on the functions
>
> Sure

>
>    -
>       - is sens a common abbreviation for sensitivity? If not, could we
>       just write it out? Code that's easier to read is helpful.
>
> "sens" is a common abbreviation in the CVODES documentation, yes. This is
why I used it.

>
>    -
>       - in the constructor, could we check the solver int values before
>       moving forward? Perhaps using check_bounded would be more
>       appropriate. I think that first if statement could be rewritten so it's
>       clearer.
>
> The very first thing I do is exactly such a if-then-else statement. If you
prefer check_bounded, we can do that.

>
>    -
>       - rather than using boost::is_same, we have template metaprograms
>       for is_const or is_const_struct which indicates whether the type is
>       constant or not.
>
> This is an important design decision from my end. So the is_same flags
within the code which of the four variants is requested, i.e. if initials
vary or not and if parameters vary or not.

>
>    -
>       - why are we passing in double values for y0 and theta instead of
>       their natural types? This would simplify the type logic in
>       cvodes_integrator.
>
> I did pass the natural types into the integator in the beginning and
decided to roll that back. The CVODES integrator does not care at all about
stan types, all it needs is to be configured correctly and all what CVODES
understands are doubles.

>
>    -
>
> Ok... before I go on, I think I need to step back and see what you're
> trying to do. I don't understand why there's so much indirection.
> integrate_ode_cvodes calls cvodes_integrator which then calls ode_model...
> to pass in values to cvodes_integrator, values are converted to double
> and things are controlled with template parameters instead of passing in
> the natural types, etc.
>
I would prefer to explain in person, but let me try to describe the
concept: The idea is that the integrate_ode_* function takes stan types and
user input which is converted into doubles (primitive types) which are
passed into the integrator. The integrator itself does not know anything
about stan's type system and as such only needs to work with doubles (BTW,
just as the odeint integration happens). The only information the
integrator needs to know is what sensitivities are to be calculated and the
cvodes_integrator object configures CVODES to generate the needed output in
an order which is later on processed with decouple_states.

Now, the next step was to abstract things a little wrt to the ODE such that
I can easily generate the input needed for CVODES which is the ODE RHS
sensitivity system. For this I need to be able to calculate the Jacobian of
the ODE RHS wrt to states and to parameters. Hence, I created an object
which wraps around the user supplied ODE which can do just that. As a side
effect of this we gain a few more nice properties of this design:

- The ode_model object can be reused to refactor the coupled_ode_system
which we currently have. This will shorten the needed code there a lot.
- In case we will add other integrators, they will also always need the
Jacobian of the ODE RHS wrt to parameters and states.
- It will allow us to easily pre-specify ODE systems within Stan which come
with analytic Jacobians. This would be of huge utility for some ODE systems
which are particularly stiff as these require often evaluation of the
Jacobians.
- This abstraction level will also allow advanced users who are willing to
code some C++ to insert analytic Jacobians for their ODE system and
eventually this may be exposed at some point to users. So this feature
makes Stan as fast as possible for ODE stuff. Moreover, there are
code-generation frameworks out there which take an ODE and generate
symbolic Jacobians in C++ code.

So what I tried to do is to tear things apart into smaller pieces which I
think will be useful in other places of the Stan code (decouple_states and
ode_model).

> I think it's in the right direction, but I'm not sure it's where it needs
> to be to replace cvode. I can spent time tomorrow trying to figure it out,
> but perhaps we should split this into multiple pull requests:
> 1. add cvodes library
> 2. add integrate_ode_cvodes (this will be a much smaller diff and we can
> make line-level comments if we split it this way)
> 3. remove integrate_ode_cvode and cvode library
>
> @wds15 <https://github.com/wds15>, thoughts?
>
We can split this pull request, sure. My suggestion:

1. add cvodes library AND remove cvode library, but leave the
integrate_ode_cvode function just as is (only the includes need to be
changed from cvode/* to cvodes/*)
2. add all of my cvodes code
3. finally remove integrate_ode_cvode

If you want to work on the code tomorrow, it could make sense to have a
video hangout briefly. So reach out to me if you think that makes sense in
order to align.

Sebastian

**Comment by syclik** [comment]
Hi Sebastian,

I'm working on the stream stuff for the next few hours, but I think I have a handle on what you're saying. I'll work through the details and get back with you. 


Daniel

**Comment by wds15** [comment]
Great. You got me thinking and I will simplify the logic a little bit, i.e. I suggest to pass the ode_model object into the cvodes_integrator will make things more straightforward.

**Comment by bob-carpenter** [comment]
This all sounds great on the design goals front.
I'm also happy to take a pass, especially if the pieces
can be pulled out of that massive pull request.  Maybe we
could just get the cvodes library in and then put the
integrator stuff in another request?

**Comment by syclik** [comment]
That was my suggestion. I can do that tonight so we have an easier diff.

**Comment by wds15** [comment]
I am on it right now... will let you now when I stop on it for today...

**Comment by wds15** [comment]
ok I am done for now. I have created an issue and a pull request for the CVODES library only. I moved all of the files from this into a new branch

feature/issue-262-CVODES-2 which I branched off from the CVODES libarary branch (and not from develop)

I hope this makes it easier to manage these pulls. tty
